File: code/calc-collaborative-similarity.php
<?php

ini_set('memory_limit', '8G');

require_once __DIR__ . '/lib/helpers.inc.php';
require_once __DIR__ . '/lib/config.inc.php';

if ($argc !== 2) die('usage: ' . $argv[0] . ' bucket');
if (!is_numeric($argv[1])) die('usage: ' . $argv[0] . ' bucket');

try
{
    $infile = sprintf('%s/data/bucket-%03d.txt', __DIR__, $argv[1]);
    $outfile = sprintf('%s/data/bucket-%03d.txt', __DIR__, $argv[1]);
    $links = unserialize(file_get_contents($infile));
    printf("N = %d\n", count($links['users']));
    printf("M = %d\n", count($links['objects']));
    $edges = 0;
    foreach ($links['users'] as $object) {
        $edges += count($object);
    }
    printf("E = %d\n", $edges);
    printf("<k> = %f\n", average_degree($links, 'users'));
    printf("<d> = %f\n", average_degree($links, 'objects'));
}
catch (PDOException $e)
{
    exit(1);
}
catch (Exception $e)
{
    exit(1);
}

printf("C_u = %s\n", network_collaborative_similarity($links, 'users', 'objects', false));
printf("s_o = %s\n", average_jaccard_similarity($links, 'objects', 'users', false));
printf("C_o = %s\n", network_collaborative_similarity($links, 'objects', 'users', false));
printf("s_u = %s\n", average_jaccard_similarity($links, 'users', 'objects', false));

File: code/calc-collaborative-similarity.py
import os
import sys
import phpserialize
from network import Network

if len(sys.argv) != 2 or not sys.argv[1].isdigit():
    exit('usage: %s bucket' % sys.argv[0])

# read data file
infile = '%s/data/bucket-%03d.txt' % (os.path.dirname(os.path.abspath(__file__)), int(sys.argv[1]))
f = open(infile)
links = phpserialize.load(f)

edges = 0
for words in links.get('users').itervalues():
    edges += len(words)

# print basic network info
print 'N = %d' % len(links.get('users'))
print 'M = %d' % len(links.get('objects'))
print 'E = %d' % edges

# calc and print extended network info
network = Network(links)
print "<k> = %f" % network.average_degree('users')
print "<d> = %f" % network.average_degree('objects')
print "C_u = %s" % (network.network_collaborative_similarity('users', 'objects', False))
print "s_o = %s" % (network.average_jaccard_similarity('objects', 'users', False))
print "C_o = %s" % (network.network_collaborative_similarity('objects', 'users', False))
print "s_u = %s" % (network.average_jaccard_similarity('users', 'objects', False))
